### DISCUSSION TITLE
How to force a page-break when convert markdown to pdf?

### DISCUSSION THREAD

**Discussion by no-vici** [Omikhleia/markdown.sile#32] [Q&A] [answered]
Huge thanks to the author of `markdown.sile` for his valuable time and hard work. 

The module does a very good job on text for converting markdown to pdf. However, I am having problem with images floating off the text-body onto the footer, I have tried to make some page-breaks by putting SILE command, \pagebreak, into markdown, it worked for a few places but not the rest. Obviously, I was stuck with "\pagebreak" displayed in my pdf.

Does anyone has any idea about this behaviour?

**Comment by Omikhleia** [comment, marked as answer]
Greetings,
Thanks for the kind feedback, it is appreciated!

To invoke SILE commands from within Markdown: as in Pandoc, you can use a fenced code block (or an inline code literal) with a raw attribute -- which, in our case would be `{=sile}`. For a page-break, typically I would suggest using the former, that is:
~~~
```{=sile}
\supereject
```
~~~

I would prefer `\supereject`, rather than `\pagebreak`, as it avoids an "underfull frame" warning (amongst other things). 

(By the way, I think SILE already has a known issue with (images or other) content _sometimes_ overlapping over the footer, but If you can share a minimal working example showing the issue, perhaps we could try to check and investigate the exact case).

**Reply by Omikhleia** [reply]
Looking closer to a related point, I moreover came to realize that the "fallbacks" for captioned tables and figures (= e.g. when not using a class or package supporting these, such my `resilient.sile` module) are not very clever with respect to page breaks... It should be be less tolerant for page breaks between the figure/table and its caption. I'll check it a bit further and open a dedicated issue. (EDITED) fixed in 53851473b631af24d9f1a008ce524c08a13015f4

**Reply by no-vici** [reply]
What a nice trick! Stupid me for not reading your well-written manual thoroughly enough, now looking back at the **Raw Block** section has help me understand the problem better and more knowledgeable about what can be done with markdown using SILE's commands.  

Taking your tip into my experiment, I have been able to manually break page at a few point and the images fit nicely. I know that SILE is not quite mature to handle the images floating smoothly, but the more I dive into SILE the more I fall in love with it, I found that what I wanted to achieve in ConTEXT Lua now becomes easier with SILE for the simplicity. I have to amid that ConTEXt is a great project and offers more features than SILE.

I will dig into your code and try to study it well.
I will carry out another experiment with `markdown.sile` with some images, and if I encounter something odd, I will come back up here and ask for solution again.
 
Thank you again for your answer.

**Reply by no-vici** [reply]
Just another small side question. 

How would I turn off chapter numbering when process markdown with `markdown.sile`?

**Reply by Omikhleia** [reply]
> How would I turn off chapter numbering when process markdown with markdown.sile?

In Markdown (well, it's a Pandoc extension actually), you can use `{-}` (or `{ .unnumbered }` after a section title do mark it as not being numbered. E.g.
```
# Some chapter {-}
```
Note that how potential subsections are then possibly numbered (if numbering on them is not also disabled the same way) is a bit beyond **markdown.sile** and rather depends on the underlying class support (e.g. I guess they'd be numbered 0.1 etc. with SILE's default "book" class; with the "resilient.book" class from [resilient.sile](https://github.com/Omikhleia/resilient.sile), they'd normally be numbered 1 etc. without trailing zeroes.).

How to change the sectioning styles globally is a another question, which can have several answers too, depending on which class is used. I have not dabbled yet into how make this easier (I don't use SILE's standard "book" class for actual projects, and the styling system in my "resilient.book" is still in infancy, with long-awaited due changes).

**Reply by no-vici** [reply]
Thank you for your quick answer. I tried `\set-counter[id=sections, numbering=false]` but it did not work. I can put the switch off `{ .unnumbered }` manually quickly into my experimenting markdown document. I have tried the book class defined in `resilient.sile` and noticed the difference in how sections and subsections are numbered.
While mentioning about the new book class in `resilient.sile`, any chances that it will allow users to insert cover pages (e.g., a full pdf cover made by IA or Inkscape), or this feature could be a taken care by the image package?
Again, thank your for you valuable time and kindness.

**Reply by Omikhleia** [reply]
Re: Cover pages: This ought to be a different question in the related repository ;) (no offense - I'm just thinking to casual readers who might have the same question as yours initially and get lost in a long exchange on unrelated topics). In brief: I have some plans for covers, but I am still experimenting with no deadline (and print books, which are my main focus, require a separate full cover...). In the meantime, an image could do the trick... But you'd want it full page (no margins), and possibly full bleed - That's where it becomes tricky...

**Reply by no-vici** [reply]
Sorry, I was too lazy to ask the previous question under the right topic, I will do that from now on. I am looking forwards to your "not near future" result ;)

**Reply by no-vici** [reply]
> Looking closer to a related point, I moreover came to realize that the "fallbacks" for captioned tables and figures (= e.g. when not using a class or package supporting these, such my `resilient.sile` module) are not very clever with respect to page breaks... It should be be less tolerant for page breaks between the figure/table and its caption. I'll check it a bit further and open a dedicated issue. (EDITED) fixed in [5385147](https://github.com/Omikhleia/markdown.sile/commit/53851473b631af24d9f1a008ce524c08a13015f4)

The last commit seems to address the footer image overlapping issue. 
Thank you!

**Comment by Omikhleia** [comment]
For the record, while the proposed escape to SILE works...
~~~
```{=sile}
\supereject
```
~~~

I also went on to add an additional "custom" extension to Markdown's horizontal dividers (#27). So a 14-dash standalone pattern (`--------------`) enforces a page break rather than showing a mere "horizontal rule". It gives a bit more flexibility to authors, IMHO, for marking a page break with this converter, but still get something visible in its place with other converters.

**Reply by no-vici** [reply]
Thank you for your time on this new extension, now users have another choice with page breaking. Adding 14 dashes could be easily done with a snippet (LuaSnip is a nice plugin). I will dig into the markdown package's code again to see what has been changed for learning. 

Much appreciated as always.

**Comment by pekkavaa** [comment]
For future readers: at least in recent versions the raw inline SILE content uses the following syntax:


    ```=sile
    \supereject
    ```

Note the missing curly braces. An alternative:

    `\supereject`{=sile}

**Reply by Omikhleia** [reply]
Yes indeed, Markdown (Pandoc-style via lunamark) and Djot (via djot-lua) differ there, in how they mark such blocks.

**Reply by Omikhleia** [reply]
This being said, the raw block/inline code trick is no longer needed for this very use case (page breaking)
 - In Markdown we recognize various types of patterns for the "horizontal divider"
 - In Djot, several attributes are available on the "thematic break" to achieve the same purpose.

(See the user manual for details.)
